### ISSUE TITLE
update to Python 3.10, pre-commit fails

### ISSUE THREAD

**Issue by nschloe** [pre-commit/pre-commit#2336]
### describe your issue

Since I updated my Ubuntu 22.04 machine to Python 3.10, pre-commit gets in my way and I can't figure out what to do. When trying to `git commit` on a Python project, pre-commit won't let me with the error message
```
[INFO] Initializing environment for https://github.com/PyCQA/isort.
[INFO] Initializing environment for https://github.com/psf/black.
[INFO] Initializing environment for https://github.com/PyCQA/flake8.
[INFO] Installing environment for https://github.com/PyCQA/isort.
[INFO] Once installed this environment will be reused.
[INFO] This may take a few minutes...
An unexpected error has occurred: AssertionError: BUG: expected environment for python to be healthy() immediately after install, please open an issue describing your environment
Check the log at /home/nschloe/.cache/pre-commit/pre-commit.log
```
See below for the log data.

I'm guessing there's an old cache somewhere, but
```
pre-commit clean
```
didn't help. Any other idea?

### pre-commit --version

pre-commit 2.18.1

### .pre-commit-config.yaml

```yaml
repos:
  - repo: https://github.com/PyCQA/isort
    rev: 5.10.1
    hooks:
      - id: isort

  - repo: https://github.com/psf/black
    rev: 22.3.0
    hooks:
      - id: black
        language_version: python3

  - repo: https://github.com/PyCQA/flake8
    rev: 4.0.1
    hooks:
      - id: flake8
```

### version information

```
pre-commit version: 2.18.1
git --version: git version 2.34.1
sys.version:
    3.10.4 (main, Apr  2 2022, 09:04:19) [GCC 11.2.0]
sys.executable: /usr/bin/python3
os.name: posix
sys.platform: linux
```

### error information

```
An unexpected error has occurred: AssertionError: BUG: expected environment for python to be healthy() immediately after install, please open an issue describing your environment
```

```
Traceback (most recent call last):
  File "/home/nschloe/.local/lib/python3.10/site-packages/pre_commit/error_handler.py", line 73, in error_handler
    yield
  File "/home/nschloe/.local/lib/python3.10/site-packages/pre_commit/main.py", line 343, in main
    return hook_impl(
  File "/home/nschloe/.local/lib/python3.10/site-packages/pre_commit/commands/hook_impl.py", line 237, in hook_impl
    return retv | run(config, store, ns)
  File "/home/nschloe/.local/lib/python3.10/site-packages/pre_commit/commands/run.py", line 414, in run
    install_hook_envs(to_install, store)
  File "/home/nschloe/.local/lib/python3.10/site-packages/pre_commit/repository.py", line 221, in install_hook_envs
    _hook_install(hook)
  File "/home/nschloe/.local/lib/python3.10/site-packages/pre_commit/repository.py", line 83, in _hook_install
    raise AssertionError(
AssertionError: BUG: expected environment for python to be healthy() immediately after install, please open an issue describing your environment
```

**Comment by asottile** [comment]
didn't you already open and then close this? #2276

**Comment by asottile** [comment]
I suspect the actual duplicate is #2299

**Comment by nschloe** [comment]
Thanks for referring me to the Ubuntu 22.04 bug. (That's one I have to deal with now.)

**Comment by asottile** [comment]
deadsnakes applies the following patch to fix this, I've proposed it to the debian maintainers as well:

```diff
diff --git a/Lib/sysconfig.py b/Lib/sysconfig.py
index a218726..f44f97c 100644
--- a/Lib/sysconfig.py
+++ b/Lib/sysconfig.py
@@ -81,6 +81,7 @@ _INSTALL_SCHEMES = {
         'data': '{base}',
         },
     }
+_INSTALL_SCHEMES['venv'] = _INSTALL_SCHEMES['posix_prefix']


 # NOTE: site.py has copy of this function.
```

**Comment by nschloe** [comment]
Nice. Where could I upvote this approach?

I've now tried to use Python from [the deadsnakes PPA](https://launchpad.net/~deadsnakes/+archive/ubuntu/ppa), but the versions are provided for past LTS version only (bionic, focal).

**Comment by asottile** [comment]
idk if there's anywhere it's publicly posted -- I have some direct contacts with the maintainers

deadsnakes typically only publishes LTS packages and since jammy isn't finalized yet I haven't built packages for it (I also typically don't repackage pythons provided by ubuntu itself, but I may make a separate PPA for jammy-3.10 with this patch applied on top of upstream)

**Comment by nschloe** [comment]
I just filed [a request on deadsnakes](https://github.com/deadsnakes/issues/issues/206), perhaps they'll accelerate jammy a little.

Any PPA with a fix would be much appreciated of course.

**Comment by asottile** [comment]
olol, it's just me 🙃 

maybe I'll make the 3.10-jammy ppa then

**Comment by nschloe** [comment]
I can also manually apply [the patch](https://github.com/pre-commit/pre-commit/issues/2336#issuecomment-1096964641) to `/usr/lib/python3.10/sysconfig.py` locally and wait for the future.

**Comment by asottile** [comment]
alright it's available: https://launchpad.net/~deadsnakes/+archive/ubuntu/python3.10-jammy

**Comment by nschloe** [comment]
Thanks!

**Comment by asottile** [comment]
I don't, personally I only care about ubuntu and the debian bug reporting process is a pain the ass so I didn't bother and just contacted the devs directly (haven't heard back)

**Comment by gaborznagy** [comment]
Sorry for spamming the issue.

May I ask why does pre-commit still failing in a venv environment, if there is a `bin/python` under the environment?
I thought the breaking change is the `posix_local` install scheme and that because of this there is not bin/python under the newly created virtual environment.

This was created with `python3 -m venv /tmp/venv` and not with `virtualenv`:
```
(venv) gaborznagy@devshell:/source$ python -m sysconfig 2>/dev/null | head -n 15
Platform: "linux-x86_64"
Python version: "3.10"
Current installation scheme: "posix_prefix"

Paths: 
        data = "/tmp/venv"
        include = "/usr/include/python3.10"
        platinclude = "/usr/include/python3.10"
        platlib = "/tmp/venv/lib/python3.10/site-packages"
        platstdlib = "/tmp/venv/lib/python3.10"
        purelib = "/tmp/venv/lib/python3.10/site-packages"
        scripts = "/tmp/venv/bin"
        stdlib = "/usr/lib/python3.10"

$python --version --version
Python 3.10.4 (main, Mar 24 2022, 13:07:27) [GCC 11.2.0]

$find /source/tests/light/ -name "*.py" -not -path "*reports*" -not -path "*tox*"  -print0 | xargs -0  pre-commit run --show-diff-on-failure --config=/source/tests/light/.pre-commit-config.yaml --files
[INFO] Installing environment for https://github.com/pre-commit/pre-commit-hooks.
[INFO] Once installed this environment will be reused.
[INFO] This may take a few minutes...
An unexpected error has occurred: AssertionError: BUG: expected environment for python to be healthy() immediately after install, please open an issue describing your environment
Check the log at /build/.cache/pre-commit/pre-commit.log
```

**Comment by asottile** [comment]
pre-commit uses `virtualenv` to create environments where tools are installed, `venv` is not an appropriate replacement as it is not reliably available

**Comment by kevinsqi** [comment]
I'm having trouble with `expected environment for python to be healthy() immediately after install` as well. I tried applying the deadsnakes patch to ubuntu 18.04.1, but that seems to fail:

```
$ sudo add-apt-repository ppa:deadsnakes/python3.10-jammy

 a small patch fixing `virtualenv` / `build` for python3.10 on ubuntu jammy

source available: https://github.com/deadsnakes/python3.10-jammy

these packages may also work on debian unstable, though no guarantees.

this package provides a drop-in `python3.10-*` package with this small patch on top of upstream:

diff --git a/Lib/sysconfig.py b/Lib/sysconfig.py
index a218726..f44f97c 100644
--- a/Lib/sysconfig.py
+++ b/Lib/sysconfig.py
@@ -81,6 +81,7 @@ _INSTALL_SCHEMES = {
         'data': '{base}',
         },
     }
+_INSTALL_SCHEMES['venv'] = _INSTALL_SCHEMES['posix_prefix']

 # NOTE: site.py has copy of this function.
 More info: https://launchpad.net/~deadsnakes/+archive/ubuntu/python3.10-jammy
Press [ENTER] to continue or Ctrl-c to cancel adding it.
Hit:1 http://us-central1.gce.archive.ubuntu.com/ubuntu bionic InRelease
Get:2 http://us-central1.gce.archive.ubuntu.com/ubuntu bionic-updates InRelease [88.7 kB]
Get:3 http://us-central1.gce.archive.ubuntu.com/ubuntu bionic-backports InRelease [74.6 kB]
Hit:4 https://packages.cloud.google.com/apt cloud-sdk InRelease
Hit:5 https://packages.cloud.google.com/apt kubernetes-xenial InRelease
Hit:6 http://security.ubuntu.com/ubuntu bionic-security InRelease
Hit:7 https://developer.download.nvidia.com/compute/cuda/repos/ubuntu1804/x86_64  InRelease
Hit:8 http://apt.postgresql.org/pub/repos/apt bionic-pgdg InRelease
Hit:9 http://ppa.launchpad.net/deadsnakes/ppa/ubuntu bionic InRelease
Ign:10 http://ppa.launchpad.net/deadsnakes/python3.10-jammy/ubuntu bionic InRelease
Hit:11 http://ppa.launchpad.net/ubuntu-toolchain-r/test/ubuntu bionic InRelease
Err:12 http://ppa.launchpad.net/deadsnakes/python3.10-jammy/ubuntu bionic Release
  404  Not Found [IP: 185.125.190.52 80]
Reading package lists... Done
E: The repository 'http://ppa.launchpad.net/deadsnakes/python3.10-jammy/ubuntu bionic Release' does not have a Release file.
N: Updating from such a repository can't be done securely, and is therefore disabled by default.
N: See apt-secure(8) manpage for repository creation and user configuration details.
```

Is there any workaround for this?

**Comment by asottile** [comment]
18.04 would be a completely different problem -- the ppa is only for jammy

please provide more information and a reproduction

**Comment by kevinsqi** [comment]
ah, I see. Here's more info: 

* On ubuntu 18.04, python 3.10.5
* I'm using a venv which was created with `python3.10 -m venv /home/kevinqi/tlaloc-deps`.
* Installed pre-commit 2.20.0 with `python -m install -r python/requirements.txt`
* I ran `pre-commit` and encountered the same `UG: expected environment for python to be healthy() immediately after install` message as the original post.


### version information

```
pre-commit version: 2.20.0
git --version: git version 2.17.1
sys.version:
    3.10.5 (main, Jun 11 2022, 16:53:29) [GCC 7.5.0]
sys.executable: /home/kevinqi/tlaloc-deps/bin/python3.10
os.name: posix
sys.platform: linux
```

### error information

```
An unexpected error has occurred: AssertionError: BUG: expected environment for python to be healthy immediately after install, please open an issue describing your environment

more info:

virtualenv python version did not match created version:
- actual version: <<error retrieving version from /home/kevinqi/.cache/pre-commit/repoi_hk8l1m/py_env-python3.10/bin/python>>
- expected version: 3.10.5.final.0

```

```
Traceback (most recent call last):
  File "/home/kevinqi/tlaloc-deps/lib/python3.10/site-packages/pre_commit/error_handler.py", line 73, in error_handler
    yield
  File "/home/kevinqi/tlaloc-deps/lib/python3.10/site-packages/pre_commit/main.py", line 386, in main
    return run(args.config, store, args)
  File "/home/kevinqi/tlaloc-deps/lib/python3.10/site-packages/pre_commit/commands/run.py", line 424, in run
    install_hook_envs(to_install, store)
  File "/home/kevinqi/tlaloc-deps/lib/python3.10/site-packages/pre_commit/repository.py", line 223, in install_hook_envs
    _hook_install(hook)
  File "/home/kevinqi/tlaloc-deps/lib/python3.10/site-packages/pre_commit/repository.py", line 84, in _hook_install
    raise AssertionError(
AssertionError: BUG: expected environment for python to be healthy immediately after install, please open an issue describing your environment

more info:

virtualenv python version did not match created version:
- actual version: <<error retrieving version from /home/kevinqi/.cache/pre-commit/repoi_hk8l1m/py_env-python3.10/bin/python>>
- expected version: 3.10.5.final.0
```

### .pre-commit-config.yaml

```
repos:
  - repo: https://github.com/psf/black
    rev: 21.7b0
    hooks:
      - id: black
        language_version: python3.6
  - repo: https://github.com/pre-commit/pre-commit-hooks
    rev: v2.4.0
    hooks:
    - id: flake8
  - repo: https://github.com/pre-commit/mirrors-eslint
    rev: 'v8.0.0-beta.0'
    hooks:
      - id: eslint
        files: \.[jt]sx?$
        additional_dependencies:
          - eslint@4.15.0
          - eslint-config-google@0.7.1
  - repo: https://github.com/sqlfluff/sqlfluff
    rev: 0.9.0
    hooks:
      - id: sqlfluff-lint
        args: ["--dialect", "postgres"]
      - id: sqlfluff-fix
        args: ["--dialect", "postgres"]
```

**Comment by asottile** [comment]
where'd you get that python from ?

**Comment by asottile** [comment]
can you run this command and show me its output?
```
find ~/.local/share/virtualenv/py_info -type f | xargs grep -l python3.10 | xargs --verbose python3 -m json.tool
```
and can you show me the version of virtualenv you are using?

**Comment by kevinsqi** [comment]
virtualenv version:
```
$ virtualenv --version
virtualenv 20.4.6 from /home/kevinqi/tlaloc-deps/lib/python3.10/site-packages/virtualenv/__init__.py
```

The command doesn't run as-is because I guess it doesn't handle multiple json inputs?

```
$ find ~/.local/share/virtualenv/py_info -type f | xargs grep -l python3.10 | xargs --verbose python3 -m json.tool
python3 -m json.tool /home/kevinqi/.local/share/virtualenv/py_info/1/062348fe29d8f1f74436f8928542fced6fab07d53c571642da1e0003e0c8f88c.json /home/kevinqi/.local/share/virtualenv/py_info/1/34a67a2f41571d663ea781008f3255d02b769f0a41b1eda57137335cda22f440.json /home/kevinqi/.local/share/virtualenv/py_info/1/30295c5bec572e859485b1ffa5e89b8b3e2022ef6e3e739c1ac40f143a557caf.json /home/kevinqi/.local/share/virtualenv/py_info/1/8a94588eda9d64d9e9a351ab8144e55b1fabf5113b54e67dd26a8c27df0381b3.json
usage: python -m json.tool [-h] [--sort-keys] [--no-ensure-ascii] [--json-lines]
                           [--indent INDENT | --tab | --no-indent | --compact]
                           [infile] [outfile]
python -m json.tool: error: unrecognized arguments: /home/kevinqi/.local/share/virtualenv/py_info/1/30295c5bec572e859485b1ffa5e89b8b3e2022ef6e3e739c1ac40f143a557caf.json /home/kevinqi/.local/share/virtualenv/py_info/1/8a94588eda9d64d9e9a351ab8144e55b1fabf5113b54e67dd26a8c27df0381b3.json
```

I can split it up and run the first two commands:

```
$ find ~/.local/share/virtualenv/py_info -type f | xargs grep -l python3.10
/home/kevinqi/.local/share/virtualenv/py_info/1/062348fe29d8f1f74436f8928542fced6fab07d53c571642da1e0003e0c8f88c.json
/home/kevinqi/.local/share/virtualenv/py_info/1/34a67a2f41571d663ea781008f3255d02b769f0a41b1eda57137335cda22f440.json
/home/kevinqi/.local/share/virtualenv/py_info/1/30295c5bec572e859485b1ffa5e89b8b3e2022ef6e3e739c1ac40f143a557caf.json
/home/kevinqi/.local/share/virtualenv/py_info/1/8a94588eda9d64d9e9a351ab8144e55b1fabf5113b54e67dd26a8c27df0381b3.json
```

If I put the first json file into json.tool:

```
python -m json.tool /home/kevinqi/.local/share/virtualenv/py_info/1/062348fe29d8f1f74436f8928542fced6fab07d53c571642da1e0003e0c8f88c.json
{
    "content": {
        "_creators": null,
        "architecture": 64,
        "base_exec_prefix": "/usr",
        "base_prefix": "/usr",
        "distutils_install": {
            "data": "",
            "headers": "include/python3.10/UNKNOWN",
            "platlib": "lib/python3.10/site-packages",
            "purelib": "lib/python3.10/site-packages",
            "scripts": "bin"
        },
        "exec_prefix": "/home/kevinqi/tlaloc-deps",
        "executable": "/home/kevinqi/tlaloc-deps/bin/python",
        "file_system_encoding": "utf-8",
        "has_venv": true,
        "implementation": "CPython",
        "max_size": 9223372036854775807,
        "original_executable": "/home/kevinqi/tlaloc-deps/bin/python",
        "os": "posix",
        "path": [
            "/home/kevinqi/tlaloc-deps/lib/python3.10/site-packages/virtualenv/discovery",
            "/home/kevinqi/tlaloc/python",
            "/usr/lib/python310.zip",
            "/usr/lib/python3.10",
            "/usr/lib/python3.10/lib-dynload",
            "/home/kevinqi/tlaloc-deps/lib/python3.10/site-packages"
        ],
        "platform": "linux",
        "prefix": "/home/kevinqi/tlaloc-deps",
        "real_prefix": null,
        "stdout_encoding": "utf-8",
        "sysconfig": {
            "makefile_filename": "/usr/lib/python3.10/config-3.10-x86_64-linux-gnu/Makefile"
        },
        "sysconfig_paths": {
            "data": "{base}",
            "include": "{installed_base}/include/python{py_version_short}{abiflags}",
            "platlib": "{platbase}/{platlibdir}/python{py_version_short}/site-packages",
            "platstdlib": "{platbase}/{platlibdir}/python{py_version_short}",
            "purelib": "{base}/lib/python{py_version_short}/site-packages",
            "scripts": "{base}/bin",
            "stdlib": "{installed_base}/{platlibdir}/python{py_version_short}"
        },
        "sysconfig_vars": {
            "PYTHONFRAMEWORK": "",
            "abiflags": "",
            "base": "/home/kevinqi/tlaloc-deps",
            "installed_base": "/usr",
            "platbase": "/home/kevinqi/tlaloc-deps",
            "platlibdir": "lib",
            "py_version_short": "3.10"
        },
        "system_executable": null,
        "system_stdlib": "/usr/lib/python3.10",
        "system_stdlib_platform": "/usr/lib/python3.10",
        "version": "3.10.5 (main, Jun 11 2022, 16:53:29) [GCC 7.5.0]",
        "version_info": {
            "major": 3,
            "micro": 5,
            "minor": 10,
            "releaselevel": "final",
            "serial": 0
        }
    },
    "path": "/home/kevinqi/tlaloc-deps/bin/python",
    "st_mtime": 1654966409.0
}
```

**Comment by asottile** [comment]
oh sorry, add a `-n1` right before the last `xargs`

**Comment by asottile** [comment]
but actually don't bother, 20.4.6 is way too old to work properly

upgrade virtualenv, purge `~/.local/share/virtualenv` and you should be good to go

**Comment by kevinsqi** [comment]
Upgrading to virtualenv==20.15.1 did the trick! This unblocks me quite a bit, really appreciate the help @asottile.

**Comment by PaleNeutron** [comment]
it doesn't work for me

```
pre-commit run --all-files
[INFO] Installing environment for https://github.com/pre-commit/pre-commit-hooks.
[INFO] Once installed this environment will be reused.
[INFO] This may take a few minutes...
An unexpected error has occurred: AssertionError: BUG: expected environment for python to be healthy immediately after install, please open an issue describing your environment

more info:

virtualenv python version did not match created version:
- actual version: <<error retrieving version from /home/john/.cache/pre-commit/repoa0vgk5yp/py_env-python3/bin/python>>
- expected version: 3.10.4.final.0

Check the log at /home/john/.cache/pre-commit/pre-commit.log
```

```
virtualenv-20.15.1
pre-commit 2.20.0
Python 3.10.4. (default python for ubuntu 22.04
```

**Comment by asottile** [comment]
@PaleNeutron because virtualenv is broken on your system -- read through this thread and the linked threads above -- there are several solutions.  if you're still stuck after researching comment here and we can help but do your due diligence first

**Comment by wooyek** [comment]
If anybody is interested there is another workaround:

```
pipx install --python $PYENV_ROOT/versions/3.9.11/bin/python3 pre-commit
```

**Comment by asottile** [comment]
@wooyek that won't help with 3.10 as that's 3.9
